### DISCUSSION TITLE
What is the proper way to retrieve Azure Tenant identifier?

### DISCUSSION THREAD

**Discussion by TechWatching** [pulumi/pulumi#9126] [Pulumi Providers]
For creating some Azure resources like an Azure Key Vault, I need to have the tenant identifier. I was wondering what was the proper way to retrieve it. I am currently doing that like this:

```csharp
public async Task<string> GetTenantId()
{
    var clientConfig = await GetClientConfig.InvokeAsync();
    var tenantId = clientConfig.TenantId;
    return tenantId;
}

var keyvault = new Vault($"kv-test-{Deployment.Instance.StackName}", new VaultArgs
{
    Properties = new VaultPropertiesArgs
    {
        TenantId = Output.Create(GetTenantId()),
        EnabledForTemplateDeployment = true,
        Sku = new Pulumi.AzureNative.KeyVault.Inputs.SkuArgs
        {
            Family = SkuFamily.A,
            Name = Pulumi.AzureNative.KeyVault.SkuName.Standard,
        },
        EnableRbacAuthorization = true,
    },
    ResourceGroupName = resourceGroup.Name
}); ;
```

Is it the correct way of doing this kind of things? As `GetClientConfig.InvokeAsync()` is an async operation and this code is in a constructor I can't make this code async and I don't want to do an `GetAwaiter().Result` or things like that.

**Comment by yuramag** [comment]
```
var tenantId = GetClientConfig.Invoke().Apply(x => x.TenantId);
```

**Comment by Zaid-Ajaj** [comment]
Hi there @yuramag and @TechWatching , both of the solutions here are fine. You can use async methods and then wrap them with `Output.Create(...)` or use the `Invoke(...).Apply(...)` variant which can be easier to work with. The async version be nice if you want to combine it with multiple async operations, then wrap their result with `Output.Create(...)`
